File: worker/app/storage.py
import boto3

from .config import Settings


class AssetStorage:
    def __init__(self, settings: Settings) -> None:
        self.bucket = settings.s3_bucket
        self.client = boto3.client(
            "s3",
            endpoint_url=settings.s3_endpoint,
            region_name=settings.s3_region,
            aws_access_key_id=settings.s3_access_key,
            aws_secret_access_key=settings.s3_secret_key,
        )

    def read(self, object_key: str) -> bytes:
        response = self.client.get_object(Bucket=self.bucket, Key=object_key)
        return response["Body"].read()

    def write(self, object_key: str, body: bytes, content_type: str) -> None:
        self.client.put_object(Bucket=self.bucket, Key=object_key, Body=body, ContentType=content_type)

    def delete(self, object_key: str) -> None:
        self.client.delete_object(Bucket=self.bucket, Key=object_key)

File: web/lib/storage.ts
import { CreateBucketCommand, DeleteObjectCommand, HeadBucketCommand, PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { getConfig } from "./config";

let client: S3Client | undefined;
function getClient(): S3Client {
  const config = getConfig();
  client ??= new S3Client({
    endpoint: config.S3_ENDPOINT,
    region: config.S3_REGION,
    forcePathStyle: true,
    credentials: { accessKeyId: config.S3_ACCESS_KEY, secretAccessKey: config.S3_SECRET_KEY },
  });
  return client;
}

let bucketPromise: Promise<void> | undefined;
async function ensureBucket(): Promise<void> {
  const config = getConfig();
  try { await getClient().send(new HeadBucketCommand({ Bucket: config.S3_BUCKET })); }
  catch { await getClient().send(new CreateBucketCommand({ Bucket: config.S3_BUCKET })); }
}

export async function storeObject(key: string, body: Buffer, contentType: string): Promise<void> {
  const config = getConfig();
  bucketPromise ??= ensureBucket();
  await bucketPromise;
  await getClient().send(new PutObjectCommand({ Bucket: config.S3_BUCKET, Key: key, Body: body, ContentType: contentType }));
}

export async function deleteObject(key: string): Promise<void> {
  const config = getConfig();
  await getClient().send(new DeleteObjectCommand({ Bucket: config.S3_BUCKET, Key: key }));
}
